File: sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int globalSum = 0;
    public int sumNumbers(TreeNode root) {
        dfs(root, 0);
        return this.globalSum;
    }

    private void dfs(TreeNode root, int sumSoFar) {
        int sum = root.val + sumSoFar * 10;
        if (root.left == null && root.right == null) {
            globalSum += sum;
            return;
        }

        if (root.left != null) {
            dfs(root.left, sum);
        }
        if (root.right != null) {
            dfs(root.right, sum);
        }

    }
}

File: 129-sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        sum = 0
        def helper(root, currVal):
            nonlocal sum
            if not root:
                return
            val = currVal * 10 + root.val
            if not root.left and not root.right:
                sum += val
                return
            helper(root.left, val)
            helper(root.right, val)
        helper(root, 0)
        return sum